cloudwatch/cloudwatch.go

```diff
@@ -289,10 +313,31 @@ func (output *OutputPlugin) AddEvent(e *Event) int {
 		return fluentbit.FLB_OK
 	}
 
-	// Step 3. Assign a log group and log stream name to the Event.
+	// Step 3. Extract the Task Metadata and the task ID if applicable.
+	if output.runningInECS && output.ecsMetadata.TaskID == "" {
+		ecsTaskMetadataEndpointV3 := os.Getenv("ECS_CONTAINER_METADATA_URI")
+		
+		metadata, err := output.getECSMetadata(ecsTaskMetadataEndpointV3)
+		if err != nil {
+			logrus.Errorf("[cloudwatch %d] Failed to get ECS Task Metadata with error: %v\n", output.PluginInstanceID, err)
+			return fluentbit.FLB_RETRY
+		}
+		output.ecsMetadata = metadata
+		arnInfo, err := arn.Parse(metadata.TaskARN)
+		if err != nil {
+			logrus.Errorf("[cloudwatch %d] Failed to parse ECS TaskARN: %s with error: %v\n", output.PluginInstanceID, metadata.TaskARN, err)
+			return fluentbit.FLB_RETRY
+		}
+		resourceID := strings.Split(arnInfo.Resource, "/")
+		taskID := resourceID[len(resourceID)-1]
+		output.ecsMetadata.TaskID = taskID
+		logrus.Debugf("[cloudwatch %d] ECS Metadata: %+v\n", output.PluginInstanceID, output.ecsMetadata)
+	}
```

Comment:
Can [most of] this chunk of new code be moved into the new `getECSMetadata` method? Then you can call the new method on 197, and set the values on line 217: `ecsMetadata: TaskMetadata{<values>},`. This approach removes a considerable amount of processing out of the `AddEvent` method.

Reply:
Thanks for the suggestion. It is a good point.

Reply:
Good changes! Thanks!